### PR TITLE
Add padding to reduce shared memory bank conflicts

### PR DESCRIPTION
**Pull request by harsh-nod**
This PR adds padding to shared memory allocations
to minimize bank conflicts.

### CODE REVIEW

File: shark_turbine/kernel/wave/promotion.py [hunk 1]
```diff
@@ -15,6 +15,25 @@
...
+    When accessing shared memory, we need to be cognizant of bank conflicts
+    that can have a significant impact on performance. One way to mitigate
+    these conflicts is by applying padding to the shared memory allocation.
+    This function applies padding of 64 bits to the shared memory allocation.
+    While this approach accomplishes the goal of reducing bank conflicts, it
+    is inefficient in terms of memory usage. A more sophisticated approach
+    would involve swizzling of the shared memory access patterns.
+    """
+    padding = 64 // dtype.bitwidth()
+    return tuple(
```
Comment (raikonenfnu) [inline shark_turbine/kernel/wave/promotion.py:31]: what if it's already aligned?

File: shark_turbine/kernel/wave/promotion.py [hunk 2]
```diff
@@ -15,6 +15,25 @@
...
+        value + padding if i == len(shape) - 1 else value
```
Comment (raikonenfnu) [inline shark_turbine/kernel/wave/promotion.py:32]: can we do something like `shape[-1] += padding`?

**Review by raikonenfnu — COMMENTED** [review]
Overall looks great just couple minor NITs/Qs.

File: shark_turbine/kernel/wave/promotion.py (hunk 1 above)
Comment (harsh-nod) [reply shark_turbine/kernel/wave/promotion.py:31]: For this approach, we just apply a blanket padding value that is independent  of whether the shape is aligned or not. We could also making this a tuning parameter to see which value gives the best performance for a given shape.

File: shark_turbine/kernel/wave/promotion.py (hunk 2 above)
Comment (harsh-nod) [reply shark_turbine/kernel/wave/promotion.py:32]: Shape is a tuple so unfortunately not.

File: shark_turbine/kernel/wave/promotion.py (hunk 1 above)
Comment (raikonenfnu) [reply shark_turbine/kernel/wave/promotion.py:31]: makes sense!

File: shark_turbine/kernel/wave/promotion.py (hunk 2 above)
Comment (raikonenfnu) [reply shark_turbine/kernel/wave/promotion.py:32]: ah I see

**Review by raikonenfnu — APPROVED** [review]

**Pull request merged**